### PR TITLE
Implement peak flow bias metric

### PR DESCRIPTION
**Pull request by SebaDro**
This pull request introduces a new metric that calculates a peak bias error in percent. The metric takes into account the shortcoming of the $\text{BiasFHV}$, which has been discussed [here](https://github.com/neuralhydrology/neuralhydrology/discussions/94).

The implementation is based on the mean peak timing metric implementation, with the difference that it does not calculates the temporal peak difference, but the bias of observed and simulated peak values in percent.

### CODE REVIEW

**Comment by kratzert** [comment]
Hi @SebaDro 

thanks for opening this PR. Just a quick note, our response might be slower over the next days as two of us are traveling/on vacation and one is still on a business trip. I'll try to look into it by Monday though!

File: neuralhydrology/evaluation/metrics.py
Comment (kratzert) [inline neuralhydrology/evaluation/metrics.py:655]: Is this fail-safe in case L 652 returns empty arrays (e.g. all obs/sim are NaN). I know we do the same above in the peak timing function (which we should probably fix) but we don't need to do the same mistake here ;) So maybe add a check that there are at least some values and otherwise return `np.nan`?

File: neuralhydrology/evaluation/metrics.py
Comment (kratzert) [inline neuralhydrology/evaluation/metrics.py:662]: I think this is open for discussion but I thought more about something like the "mean absolute peak flow bias (in percent)", rather than the overall peak bias.

IIuc, in your equation, over- and underestimates of different peaks could cancel each other out (because of the missing `np.abs()`, leading potentially to a perfect peak-bias score even though there are errors. Also, by summing over all errors first (and all obs), we would neglect that absolute errors on different magnitudes are in different value ranges.

E.g. imagine we have two peaks, one with obs=2 and one with obs=200. For the first peak, imagine sim=1 and for the second sim=180. According to your equation, the peak_flow_bias would be ~10% (21/202 * 100). If you would compute the percentage bias on each peak first, and then average across peaks, you would get (1/2 + 20/200)/2 * 100 = 30%.

Since this PR is proposing a new metric, there is no right or wrong, but intuitively (also considering the other metrics here, like the peak timing metric), I would find the "mean peak bias in percent" clearer. @gauchm @danklotz opinions?

**Comment by kratzert** [comment]
Thanks again for your PR. I think overall the function looks great and I mainly have one point (the second comment) about "What does this metric compute". I would love to hear @gauchm and/or @danklotz opinion on that.

File: neuralhydrology/evaluation/metrics.py
Comment (SebaDro) [reply neuralhydrology/evaluation/metrics.py:655]: Actually, I tested the implementation with some saboteurs, which cause different failures. In your described case, the routine shows a runtime warning but will not fail with an error. It also ends up with an `np.nan` as a result. However, I agree that a fail-fast would be ok here, since it improves readability of the implementation by explicitly returning `np.nan` in failing cases. I will address it with a dedicated commit.

File: neuralhydrology/evaluation/metrics.py
Comment (SebaDro) [reply neuralhydrology/evaluation/metrics.py:662]: The equation is inspired by the $\text{BiasFHV}$ equation, which also does not take into account absolute differences in simulated and observed flows. I see that there is a shortcoming, resulting from the coincident of postive and negative differences, which eliminates each other. However, if we take the absolute differences for calculation, the resulting metric does not say anaything about over- or underestimates of the peaks.

Regarding the question about considering the overall or mean peak bias, I'm not sure which approach yields a more valuable accuracy metric. Sure, taking the mean of all differences is much clearer and each peak flow difference is weighted equally. However, differences in very high peaks have a greater impact on the result than differences in lower peaks. My feeling is, that a difference of e.g. 10 m³/s in a very high peak is less relevant than the same difference in very low peaks. Though, normalizing the difference by the sum of all observed flows, such as in the $\text{BiasFHV}$, adresses this issue.

So, as you already mentioned, a discussion about this new metric would be very helpful to clear up its understanding.

**Comment by SebaDro** [comment]
Thanks for your review. I added some notes as inline comments in the conversations. Looking forward to some discussions about the metric itself and its implementation.

File: neuralhydrology/evaluation/metrics.py
Comment (gauchm) [reply neuralhydrology/evaluation/metrics.py:670]: To be honest, I don't have much of an opinion on this issue - sorry. For me both the abs and the non-abs approach are totally fine.
I think I haven't fully understood the argument about overall vs. mean peak bias. Isn't the impact of the same difference for a high peak _lower_ (and not higher) than for a low peak (since Freddy wants to divide by the corresponding obs)?

File: neuralhydrology/evaluation/metrics.py
Comment (gauchm) [reply neuralhydrology/evaluation/metrics.py:659]: I'm in favor of an explicit empty array edge case, given that I think the `find_peaks` function does not explicitly define its behavior for this case (in their docs).

File: neuralhydrology/evaluation/metrics.py
Comment (SebaDro) [reply neuralhydrology/evaluation/metrics.py:670]: >Isn't the impact of the same difference for a high peak lower (and not higher) than for a low peak (since Freddy wants to divide by the corresponding obs)?

Yes, absolutely. Actually, that is what I meant. The argument regarding higher and lower impact relates to the mean of all differences. Taking the mean will score all differences equally. So differences in high peaks, which are likely to be also high, will impact the mean more than differences in lower peaks, which are likely to be lower. But instead, as you also mention, differences in lower peaks should have an higher impact than the same differences in higher peaks.

Comment (danklotz) [reply neuralhydrology/evaluation/metrics.py:670]: As Martin I also have no strong feelings about this. Here are my thoughts: 

If we want to call the metric "xyz bias" then I would for using the version without the absolute. My reasoning is that it then does more closely correspond to what math people would expect from a bias (in the sence that we check if the expectations do correspond). 

From a diagnostic standpoint, however, I like the version with the absolute better. As you say, we loose information about wheter we over- or underestimate (which imo is also not well captured by the current bias), but we would be much sharper with respect to the error magnitudes. 

Since over- and underestimation is important we might still want to recuperate that info at some point (maybe yet another metric?).

Comment (kratzert) [reply neuralhydrology/evaluation/metrics.py:670]: So to summarize, 3 out of 4 think the metric as is is fine? Then I vote for merging the metric into master. We can always add other metrics and/or modify the metric later. 

Thanks again @SebaDro for your effort here. I will wait for feedback until tomorrow, otherwise I will merge it. And sorry again for my delayed replies but I am currently on vacation and only have a quick look at the internet in the evening.

Comment (danklotz) [reply neuralhydrology/evaluation/metrics.py:670]: I think it is a 2 ( @kratzert + @danklotz ) vs 2 ( @gauchm + @SebaDro ) currently. 

Just to make my position clear: I prefer the version of @kratzert (The one with the `np.abs`). The reason for that is that it seems to me the better diagnostic tool. The other version can compensate errors (a) between peaks and (b) between basins. We loose some info about over and under estimation, but imo gain more.

I just would like to call it differently; say *relative mean absolute peak error*.

File: neuralhydrology/evaluation/metrics.py
Comment (SebaDro) [reply neuralhydrology/evaluation/metrics.py:659]: In the meanwhile, I've already created another commit that addressed this issue. I think, that's the reason why ;-)

File: neuralhydrology/evaluation/metrics.py
Comment (SebaDro) [reply neuralhydrology/evaluation/metrics.py:670]: Actually, I'd be fine with both versions. Although I introduced the non-abs version, I also think that taking the absolute error gives more information about the model performance in peak regions. So, it's totally ok, if we vote for that metric. 

Though, there is still one decision pending regarding taking the mean absolute error or the overall percent bias? As far as I can see, there are three different implementation and metric name options:
1. Mean absolute peak error (MAE_peak):
$\text{MAE}\_\text{peak} =\frac{1}{P} \sum_{p=1}|Q_{s,p} - Q_{o,p}|$
```python
mae_peak = np.sum(np.abs(sim - obs)) / peaks.size()
```

2. Absolute percentage peak error (APE_peak):
$\text{APE}\_\text{peak} = \frac{\sum_{p=1}|Q_{s,p} - Q_{o,p}|}{\sum_{p=1}Q_{o,p}} \times 100$
```
ape_peak = np.sum(np.abs(sim - obs)) / np.sum(obs) * 100
```

3. Mean absolute percentage peak error (MAPE_peak):
$\text{MAPE}\_\text{peak} =  \frac{1}{P}\sum_{p=1}\frac{|Q_{s,p} - Q_{o,p}|}{Q_{o,p}}\times 100$
```
mape_peak = np.sum(np.abs(sim - obs) / np.sum(obs)) / peaks.size() * 100
```

I'd vote for option three, since it is a well known metric for percentage prediction errors in statistics. There also exists a [scikit-learn implementation](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.mean_absolute_percentage_error.html) for it. So, I'd assume that there should be a common understanding of this metric. 

Just give a vote for one of the options and I will update the pull request in accordance to the voting result.

Comment (danklotz) [reply neuralhydrology/evaluation/metrics.py:670]: Sorry if I am annoying, but from what I gather MAPE seems to be defined slightly different (see [a](https://en.m.wikipedia.org/wiki/Mean_absolute_percentage_error),[b](https://github.com/scikit-learn/scikit-learn/blob/36958fb24/sklearn/metrics/_regression.py#L296)). Apart from that I am fine with 2, 3, or other MAPE versions ☺️

Comment (SebaDro) [reply neuralhydrology/evaluation/metrics.py:670]: Oh yes, thanks. In fact, the above shown MAPE implementation does not fit to the formula. Actually (assuming, observations are always positive), it should be:
```
mape_peak = np.sum(np.abs(sim - obs) / obs) / peaks.size() * 100
```
 But to fit the original formula, we also could write :
```
mape_peak = np.sum(np.abs((sim - obs) / obs)) / peaks.size() * 100
```
The formula then would be $\text{MAPE}\_\text{peak} =  \frac{1}{P}\sum_{p=1}|\frac{Q_{s,p} - Q_{o,p}}{Q_{o,p}}|\times 100$, with $\text{P}$ as the number of peaks considered.

Comment (kratzert) [reply neuralhydrology/evaluation/metrics.py:670]: +1 looks good to me.

Comment (gauchm) [reply neuralhydrology/evaluation/metrics.py:670]: +1 from me as well. I'd say let's go ahead with this version?

Comment (danklotz) [reply neuralhydrology/evaluation/metrics.py:670]: +1

Comment (SebaDro) [reply neuralhydrology/evaluation/metrics.py:670]: Great 👍. I'll come with an updating commit as soon as possible.

**Review by kratzert — APPROVED** [review]
LGTM, thanks @SebaDro for this PR and congratulations on your first PR to NeuralHydrology. Hopefully more to come in the future ;) 

I'll go ahead and merge the PR into master.

**Pull request merged**
